.github/workflows/build.yml

```
@@ -25,6 +25,10 @@ on:
 permissions:
   contents: read
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
```

Have you verified if `github.ref` is always defined based on the triggers we use for these workflows?

The docs say
> This is only set if a branch or tag is available for the event type.

If there are cases where it's undefined, it might end up cancelling a bunch of other workflows that were triggered in the same way, so we have to be careful.  The syntax supports fallback values with `${{ github.ref || fallback_value}}`

---

I am using `${{ github.head_ref || github.run_id }}`, it works quite well for a wide range of different workflows.

---

> I am using ${{ github.head_ref || github.run_id }}

This is also what is being suggested at https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-a-fallback-value
```suggestion
  group: ${{ github.head_ref || github.run_id }}
```
This adds a fallback on `github.run_id` to the original group:
```suggestion
  group: ${{ github.workflow }}-${{ github.ref || github.run_id }}
```
This adds a fallback on `github.run_id` and uses `github.head_ref` instead of `github.ref` (note that `head_ref` "is only available when the event that triggers a workflow run is either `pull_request` or `pull_request_target`"):
```suggestion
  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
```
Pick one :)

---

Thanks both, I picked the last one to match mypy:

https://github.com/python/mypy/blob/589ad1c17eeb220a41ba41425b61b8593f8bc42d/.github/workflows/test.yml#L19